cloudinary-image-management-and-manipulation-in-the-cloud-cdn/php/class-sync.php

```diff
@@ -119,66 +163,654 @@ public function is_synced( $post_id ) {
+	 *
+	 * @return bool
+	 */
+	public function is_pending( $attachment_id ) {
+		// Check if it's not already in the to sync array.
+		if ( ! in_array( $attachment_id, $this->to_sync, true ) ) {
+			$is_pending = $this->managers['media']->get_post_meta( $attachment_id, Sync::META_KEYS['pending'], true );
+			if ( empty( $is_pending ) || $is_pending < time() - 5 * 60 ) {
+				// No need to delete pending meta, since it will be updated with the new timestamp anyway.
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	/**
+	 * Add an attachment ID to the to_sync array.
+	 *
+	 * @param int $attachment_id The attachment ID to add.
+	 */
+	public function add_to_sync( $attachment_id ) {
+		if ( ! in_array( $attachment_id, $this->to_sync, true ) ) {
+			// Flag image as pending to prevent duplicate upload.
+			$this->managers['media']->update_post_meta( $attachment_id, Sync::META_KEYS['pending'], time() );
+			$this->to_sync[] = $attachment_id;
+		}
+	}
+
+	/**
+	 * Update a part of the signature based on type.
+	 *
+	 * @param int    $attachment_id The attachment ID.
+	 * @param string $type          The type of sync.
+	 */
+	public function update_signature( $attachment_id, $type ) {
+
+		$expecting           = $this->generate_signature( $attachment_id );
+		$current_sync_method = $this->sync_base_struct[ $type ]['sync']['callback'];
+
+		// Go over all other types that share the same sync method and include them here.
+		foreach ( $this->sync_base_struct as $sync_type => $struct ) {
+			if ( $struct['sync']['callback'] === $current_sync_method ) {
+				$this->set_signature_item( $attachment_id, $sync_type, $expecting[ $sync_type ] );
+			}
+		}
+
+	}
+
+	/**
+	 * Set an item to the signature set.
+	 *
+	 * @param int    $attachment_id The attachment ID.
+	 * @param string $type          The sync type.
+	 * @param null   $value
+	 */
+	public function set_signature_item( $attachment_id, $type, $value = null ) {
+
+		// Get the core meta.
+		$meta = wp_get_attachment_metadata( $attachment_id, true );
+
+		// Set the specific value.
+		if ( is_null( $value ) ) {
+			// Generate a new value based on generator.
+			$new_value = call_user_func( $this->sync_base_struct[ $type ]['generate'], $attachment_id );
+			if ( is_array( $new_value ) ) {
+				$new_value = wp_json_encode( $new_value );
+			}
+			$value = md5( $new_value );
+		}
+		$meta[ Sync::META_KEYS['cloudinary'] ][ Sync::META_KEYS['signature'] ][ $type ] = $value;
+		wp_update_attachment_metadata( $attachment_id, $meta );
+	}
+
+	/**
+	 * Initialize the background sync on requested images needing to be synced.
+	 */
+	public function init_background_upload() {
+		if ( ! empty( $this->to_sync ) ) {
+
+			$threads    = $this->managers['push']->queue->threads;
+			$chunk_size = ceil( count( $this->to_sync ) / count( $threads ) ); // Max of 3 threads to prevent server overload.
+			$chunks     = array_chunk( $this->to_sync, $chunk_size );
+			$token      = uniqid();
```

Comment:
As said before, we shouldn't rely on this for true uniqueness.

Reply:
@dugajean - This is not really for uniqueness, but simply to have a key that references to a transient. Which only last 120 seconds. It also gets the thread ID attached to keep them separated.